_posts/2025-12-15-quarkus-feature-flags.asciidoc

```diff
@@ -0,0 +1,264 @@
+layout: post
+title: 'Quarkus feature flags'
+date: 2025-12-15
+tags: extension
+synopsis: 'We introduce a new Quarkiverse project - a lightweight and extensible feature flags extension.'
+author: mkouba
+---
+
+Feature flags are a proven and popular technique.
+In essence, a feature flag makes it possible to turn on-off and/or configure a specific functionality in your application.
+It is also referred to as toggles or switches.
+There are many types of feature flags.
+You may have heard of a "killer switch" that simply disables a problematic feature instantly.
+Feature flags can be used to implement "gradual rollout" to deliver new features gradually to small groups of users (aka beta testers).
+Permission flags are used to control access for different users.
+And so on.
+However, that's not the subject of this blogpost.
+We would like to introduce [Quarkus Feature Flags](https://github.com/quarkiverse/quarkus-flags/) - a Quarkiverse project that aims to provide a _lightweight_ and _extensible_ feature flags Quarkus extension.
+
+More specifically, it provides:
+
+- A blocking/non-blocking API to access feature flags.
+- A non-blocking SPI to provide flags and externalize the computation of a flag value.
+- Several built-in flag providers:
+   - Quarkus config can be used to define feature flags,
+   - An in-memory flag repository which is useful for testing and dynamic registration of flags.
+- *Hibernate ORM* module, where feature flags are mapped from an annotated entity and are automatically loaded from the database.
+- *Security* module, so that it's possible to evaluate flags based on the current `SecurityIdentity`.
+- *Qute* module, so that it's possible to use the flags directly in templates.
+- *CRON* module with a flag evaluator that matches a specific CRON expression.
+
+Note that the goal of this extension is not to replace robust solutions such as [OpenFeature](https://openfeature.dev/) and [Unleash](https://www.getunleash.io/).
+Instead, we would like to offer a flexible option that integrates well with other parts of the Quarkus ecosystem.
+
+== Flag
+
+In this extension, a feature flag is represented by the `io.quarkiverse.flags.Flag` interface.
+It refers to a specific feature with a string identifier and provides several convenient methods to compute the _current value_.
+The value of a feature flag can be represented as `boolean`, `string` or `integer`.
+There can be only one flag for a given feature at a given time.
+A flag can also expose metadata which that can be leveraged in the SPI.
+
+== Simple example
+
+Let's start simple.
+We will create a flag for a feature called `my-feature-alpha` in the `application.properties` file.
+
+[source,properties]
+----
+quarkus.flags.runtime."my-feature-alpha".value=true <1>
+----
+<1> Define a _runtime_ flag for feature `my-feature-alpha` with initial value `true`.
+
+TIP: The runtime feature flags can be overriden at runtime, e.g. using a system property or an environment variable.
```

Comment:
Are there build-time flags? Not that they would make any sense, they would be just configuration, but this comment makes me ask :-)

Reply:
Yes, there are. And they actually make sense if you need to define a build time fixed flag. Because that's not possible with regular Quarkus config (from app POV).

Reply:
Well, I guess if you truly embrace the Quarkus build time architecture, build-time feature flags may make sense in order to build multiple variants of the application where each behaves slightly differently... I think we've mostly seen demands for the opposite though, but fair enough. I think you could mention them here at least.